add `utc_from_now/1` and `utc_ago/1` to all calendar types

tfiedlerdejanze opened the pull request.

zolrath commented on lib/elixir/lib/calendar/datetime.ex:
```diff
@@ -1829,6 +1829,46 @@ defmodule DateTime do
+
+  ## Examples
+
+      iex> ndt = DateTime.utc_ago(day: 1)
```
```suggestion
      iex> ndt = DateTime.utc_from_now(day: 1)
```

zolrath commented on lib/elixir/lib/calendar/datetime.ex:
```diff
@@ -1829,6 +1829,46 @@ defmodule DateTime do
+      iex> ndt = DateTime.utc_ago(day: 1)
+      iex> ndt == DateTime.add(DateTime.utc_now(), 1, :day)
+
+      iex> ndt = DateTime.utc_ago(month: 1)
```
```suggestion
      iex> ndt = DateTime.utc_from_now(month: 1)
```

zolrath commented on lib/elixir/lib/calendar/naive_datetime.ex:
```diff
@@ -650,6 +650,46 @@ defmodule NaiveDateTime do
+
+  ## Examples
+
+      iex> ndt = NaiveDateTime.utc_ago(day: 1)
```
```suggestion
      iex> ndt = NaiveDateTime.utc_from_now(day: 1)
```

zolrath commented on lib/elixir/lib/calendar/naive_datetime.ex:
```diff
@@ -650,6 +650,46 @@ defmodule NaiveDateTime do
+      iex> ndt = NaiveDateTime.utc_ago(day: 1)
+      iex> ndt == NaiveDateTime.add(NaiveDateTime.utc_now(), 1, :day)
+
+      iex> ndt = NaiveDateTime.utc_ago(month: 1)
```
```suggestion
      iex> ndt = NaiveDateTime.utc_from_now(month: 1)
```

zolrath commented on lib/elixir/lib/calendar/time.ex:
```diff
@@ -605,6 +605,50 @@ defmodule Time do
+
+  ## Examples
+
+      iex> time = Time.utc_ago(hour: 1)
```
```suggestion
      iex> time = Time.utc_from_now(hour: 1)
```

zolrath commented on lib/elixir/lib/calendar/time.ex:
```diff
@@ -605,6 +605,50 @@ defmodule Time do
+      iex> time = Time.utc_ago(hour: 1)
+      iex> time == Time.add(Time.utc_now(), 1, :hour)
+
+      iex> time = Time.utc_ago(minute: 30)
```
```suggestion
      iex> time = Time.utc_from_now(minute: 30)
```

zolrath reviewed (commented):
Just a few copy/paste issues in the examples!

tfiedlerdejanze commented:
> Just a few copy/paste issues in the examples!

Thanks! actually turns out the examples aren't being properly interpreted as the CI did actually pass :/ Either way, happy for some suggestions to better illustrate how the function works in the examples.

I changed and moved some of the examples to the test files to make sure the behaviour is properly tested.

wojtekmach commented on lib/elixir/lib/calendar/datetime.ex:
```diff
@@ -1829,6 +1829,50 @@ defmodule DateTime do
+  """
+  @doc since: "1.17.0"
+  @spec utc_ago(Duration.t() | [Duration.unit_pair()]) :: t
+  def utc_ago(%Duration{} = duration) do
```
there is no rush but curious if we should add a `precision \\ :microsecond` to mimic `DateTime.utc_now(:second)`.

it would be a bit awkward with keyword argument though:

```elixir
DateTime.utc_ago([hour: 42], :second)
```

so maybe it’d be:

```elixir
DateTime.utc_ago(hour: 42, precision: :second)
```

and so probably no Calendar.duration() type after all. 😅

tfiedlerdejanze replied:
We could be adding both arity `/1` and `/2` versions

```
DateTime.utc_ago(%Duration{hour: 42}, :second)
```

or

```
DateTime.utc_ago(hour: 42, precision: :second)
```

tfiedlerdejanze commented on lib/elixir/lib/calendar/date.ex:
```diff
@@ -810,6 +810,50 @@ defmodule Date do
+
+  """
+  @doc since: "1.17.0"
+  @spec utc_ago(Duration.t() | [Duration.unit_pair()]) :: t
```
I dont think it was but it seems like a reasonable addition! Might become easier to reason about once we added more functions supporting Durations

josevalim replied:
I think this type is good and I would add it to Duration itself. We should probably make it so `Duration.new!/1` also accepts an existing duration and just return it. This way we can simplify functions such as `shift` to just call `Duration.new!`? This can be done separately from this PR. :)

tfiedlerdejanze replied:
here it is: https://github.com/elixir-lang/elixir/pull/13476 thanks for the input!

josevalim commented:
Thank you for the PR @tfiedlerdejanze!

Looking at the code makes a huge difference and, given @wojtekmach's suggestion that we also want to pass the precision as argument, I am not sure if this function offers much compared to just chaining them together. I think we should wait until we are sure the benefits here are worth it.

tfiedlerdejanze commented:
Thanks @josevalim! I agree, looking at the code, there is actually very little lift we're doing for what's already achievable in a single line of code
